src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,425 @@
+                    break;
+                }
+
+                case qc::OpType::Phase: {
+                    const auto phase = PiRational(op->getParameter().front());
+                    addCphase(diag, phase, ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::T: {
+                    addCphase(diag, PiRational(1, 4), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::S: {
+                    addCphase(diag, PiRational(1, 2), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::Tdag: {
+                    addCphase(diag, PiRational(-1, 4), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::Sdag: {
+                    addCphase(diag, PiRational(-1, 2), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                default: {
+                    throw ZXException("Unsupported Controlled Operation: " +
+                                      qc::toString(op->getType()));
+                }
+            }
+        } else if (op->getNcontrols() == 2) {
+            Qubit       ctrl_0 = 0;
+            Qubit       ctrl_1 = 0;
+            const Qubit target = p.at(op->getTargets().front());
+            int         i      = 0;
+            for (auto& ctrl: op->getControls()) {
+                if (i++ == 0)
+                    ctrl_0 = p.at(ctrl.qubit);
+                else
+                    ctrl_1 = p.at(ctrl.qubit);
+            }
+            switch (op->getType()) {
+                case qc::OpType::X: {
+                    addCcx(diag, ctrl_0, ctrl_1, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::Z: {
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    addCcx(diag, ctrl_0, ctrl_1, target, qubit_vertices);
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    break;
+                }
+                default: {
+                    throw ZXException("Unsupported Multi-control operation: " +
+                                      qc::toString(op->getType()));
+                    break;
+                }
+            }
+        } else {
+            throw ZXException("Unsupported Multi-control operation (" + std::to_string(op->getNcontrols()) + " ctrls)" + qc::toString(op->getType()));
+        }
+        return it + 1;
+    }
+
+    ZXDiagram FunctionalityConstruction::buildFunctionality(const qc::QuantumComputation* qc) {
+        ZXDiagram           diag(qc->getNqubits());
+        std::vector<Vertex> qubit_vertices(qc->getNqubits());
+        for (size_t i = 0; i < qc->getNqubits(); ++i) {
```

Comment:
```suggestion
        for (std::size_t i = 0; i < qc->getNqubits(); ++i) {
```
Might hold for some more places.